### DISCUSSION TITLE
Is it possible that BBolt cannot delete a key when the storage has no space left?

### DISCUSSION THREAD

**Discussion by andrzej-stencel** [etcd-io/bbolt#685] [Q&A] [answered]
Hi,

I'm dealing with an issue in the OpenTelemetry Collector that uses BBolt for the exporter queues (using the [File Storage extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.93.0/extension/storage/filestorage/client.go#L51)). It seems that I've hit a scenario when the exporter wasn't able to remove an item from the queue after the underlying storage device had no space left. The actual operation was `bucket.Delete` in a transaction: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.93.0/extension/storage/filestorage/client.go#L118. It was called from here: https://github.com/open-telemetry/opentelemetry-collector/blob/v0.92.0/exporter/exporterhelper/internal/persistent_queue.go#L489.

I admit I know next to nothing about BBolt. My thinking is that due to the transactional/versioning nature of BBolt, it might need to reserve a new page to create a new version of the node(s) that don't contain the deleted data. Does this make sense?

**Comment by ahrtr** [comment]
I do not get what's the exact issue you are running into. Do you mean that `bucket.Delete` returned an error or the `tx.Commit` somehow failed?

**Reply by andrzej-stencel** [reply]
Looking at [bucket.Delete](https://github.com/etcd-io/bbolt/blob/v1.3.8/bucket.go#L310) method I don't think it's possible for this error to come from this method. The error was `write /var/lib/storage/otc/exporter_otlphttp__logs: no space left on device`.

I got the error from the [bbolt#DB.Update](https://pkg.go.dev/go.etcd.io/bbolt#DB.Update) method [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.92.0/extension/storage/filestorage/client.go#L133). I'm guessing it must have come from the [tx.Commit](https://github.com/etcd-io/bbolt/blob/v1.3.8/tx.go#L142) method.

**Reply by ahrtr** [reply]
There are two possible sources of the error:
- The callback function [batch](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e833bc5e1ace83593225d714f16f8f221b4c46ad/extension/storage/filestorage/client.go#L96). Please double confirm this
- The tx.Commit

Please provide the detailed error you received.

**Reply by ahrtr** [reply]
FYI. bbolt supports printing log starting from the 1.4.0 (the main branch for now), you can also try the main branch so that you can see more debug messages. Please refer to https://github.com/etcd-io/etcd/pull/17237

**Reply by andrzej-stencel** [reply]
> Please provide the detailed error you received.

The error was `write /var/lib/storage/otc/exporter_otlphttp__logs: no space left on device`.

**Reply by ahrtr** [reply]
> The error was write /var/lib/storage/otc/exporter_otlphttp__logs: no space left on device.


The error was returned by OS (syscall), and it's clear that there is no space left. Each time when committing a bolt transaction, it will try to reuse the existing free pages, and allocate more disk space (pages) only when the existing free pages can't meet the requirement. 

So this isn't an issue. Can we close this session?

**Reply by andrzej-stencel** [reply]
I didn't say it was an issue 😄 I asked a specific question: Is it possible that BBolt fails to remove a key if there's not enough storage available?

I'm assuming from your replies that the answer to the question is "Yes". That was also my initial assumption, but I wanted to confirm this. Thank you!

**Reply by ahrtr** [reply, marked as answer]
> I didn't say it was an issue 😄 I asked a specific question: Is it possible that BBolt fails to remove a key if there's not enough storage available?

Oh, sorry for the confusion. The answer is indeed "YES". The same reason as I mentioned in https://github.com/etcd-io/bbolt/discussions/685#discussioncomment-8317526. Each time when updating (no matter it's writing or deleting), bbolt will always write the updated data into new pages, the previous pages will keep unchanged to ensure that the db file isn't corrupted if the transaction runs into any issue.

Please refer to https://github.com/ahrtr/etcd-issues/blob/master/docs/cncf_storage_tag_etcd.md#storage-boltdb-feature

**Reply by andrzej-stencel** [reply]
Awesome, thank you so much. The linked information is very useful too. I believe it would be great to add these sections to the BBolt documentation (the README file):
- [Storage: BoltDB](https://github.com/ahrtr/etcd-issues/blob/master/docs/cncf_storage_tag_etcd.md#storage-boltdb)
- [Storage: BoltDB feature](https://github.com/ahrtr/etcd-issues/blob/master/docs/cncf_storage_tag_etcd.md#storage-boltdb-feature)
- Perhaps this one too: [Storage: tools to analyze BoltDB file](https://github.com/ahrtr/etcd-issues/blob/master/docs/cncf_storage_tag_etcd.md#storage-tools-to-analyze-boltdb-file)
